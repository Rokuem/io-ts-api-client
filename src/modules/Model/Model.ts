import * as t from 'io-ts';
import reporter from 'io-ts-reporters';
import { Either, isRight } from 'fp-ts/lib/Either';
import { TypedEmitter } from '../TypedEmitter/TypedEmmiter';
import { addedDiff } from 'deep-object-diff';
import { ValidationOptions } from '../types';
import { deepMerge } from '../../helpers/deepMerge';

/**
 * Simple io-ts model with a validator. This is used to safeguard against API changes and to provide a easy solution for type checking it.
 */
export class Model<Schema extends t.Any = any, SampleKey extends string = never, Samples extends t.TypeOf<Schema> = t.TypeOf<any>> {
  /**
   * Emitter for modal events.
   */
  public static emitter = new TypedEmitter<{
    'before-validation': (model: string) => void;
    'validation-success': (model: string) => void;
    'validation-error': (model: string, error: Error) => void;
    'after-validation': (model: string) => void;
    'extra-keys-detected': (model: string, error: Error) => void;
  }>();

  /**
   * a name to call the model during logs and errors.
   */
  public name: string;

  /**
   * @deprecated - use model.schema instead.
   */
  public base: Schema;

   /**
   * IO-TS interface used for the model.
   *
   * you can use this to extend other interfaces.
   */
  public schema!: Schema;

  /**
   * Ts interface of the model.
   * 
   * !Warn: Do not use this as an value.
   * @since 0.7.4
   */
  public tsInterface!: t.TypeOf<Schema>

  public samples!: Record<SampleKey, Samples>;

  public constructor(config: {
    /**
     * The name of the model.
     */
    name: string;
    /**
     * The io-ts model.
     */
    schema: Schema;
    samples?: Record<SampleKey, Samples>;
  }) {
    this.name = config.name;
    this.base = config.schema;
    this.schema = config.schema;
    this.samples = config.samples || {} as any;
  }

  /**
   * If not an empty string, this model belongs to an operation.
   */
  public operation = '';

  /**
   * Checks if there are any extra keys in the response received.
   */
  private checkForExtraKeys(options: {
    result: Either<t.Errors, any>;
    target: any;
    debug: boolean;
    throwErrors: boolean;
  }) {
    const { result, target } = options;

    if (!('right' in result) && options.debug) {
      console.warn('Extra fields check skipped due to failed validation');
      return;
    }

    if ('props' in this.base && 'right' in result) {
      const filteredDecode = t.exact(this.base as any).decode(target);
      const resultJson = JSON.stringify(result.right);

      if (!('right' in filteredDecode)) {
        throw new Error('Something went wrong while checking for extra keys');
      }

      const filtered = filteredDecode.right;
      const filteredJson = JSON.stringify(filtered);

      if (filteredJson !== resultJson) {
        const diff = addedDiff(filtered, result.right);
        const errorMessage = `${
          this.operation ? `[${this.operation}] ` : ''
        }Detected extra properties in model "${this.name}": ${JSON.stringify(
          diff,
          null,
          '  '
        )}`;

        Model.emitter.emit('extra-keys-detected', this.name, new Error(errorMessage));

        if (options.throwErrors) {
          throw new Error(errorMessage);
        } else if (options.debug) {
          console.error(errorMessage);
        }
      }

      return filtered;
    }
  }

  private handleInvalidResult(options: {
    result: Either<t.Errors, any>;
    target: any;
    throwErrors: boolean;
    debug: boolean;
  }) {
    const validationErrors = reporter.report(options.result);

    for (const error of validationErrors) {
      const validationMessage = `ERROR - ${
        this.operation ? `[${this.operation}]` : ''
      }[${this.name}] VALIDATION FAILED: ${error} at: \n ${JSON.stringify(
        options.target,
        null,
        2
      )}`;

      Model.emitter.emit('validation-error', this.name, new Error(validationMessage));

      if (options.debug) {
        console.error(validationMessage);
      }

      if (options.throwErrors) {
        throw new Error(validationMessage);
      }
    }
  }

  private handleValidResult(
    target: any,
    options: ValidationOptions
  ) {
    if (options.debug) {
      console.log(
        `SUCCESS - ${this.operation ? `[${this.operation}]` : ''}[${
          this.name
        }]: Target is valid!`,
        {
          target,
        }
      );
    }

    Model.emitter.emit('validation-success', this.name);
  }

  /**
   * Validates the model. Throws error if failed in strict mode.
   */
  public validate(
    target: any,
    options: ValidationOptions
  ) {
    Model.emitter.emit('before-validation', this.name);

    let result = this.base.decode(target);
    const isValid = isRight(result);

    if (options.strictTypes) {
      const filtered = this.checkForExtraKeys({
        result,
        target,
        debug: options.debug,
        throwErrors: options.throwErrors,
      });

      if (filtered) {
        result = {
          ...result,
          right: filtered,
        } as any;
      }
    }

    if (!isValid) {
      try {
        this.handleInvalidResult({
          target,
          result,
          debug: !!options.debug,
          throwErrors: !!options.throwErrors,
        });
      } catch (e) {
        Model.emitter.emit('after-validation', this.name);
        throw e;
      }
    } else {
      this.handleValidResult(target, options);
    }

    Model.emitter.emit('after-validation', this.name);

    return 'right' in result ? result.right : target;
  }

  /**
   * Asserts that target is valid for this model.
   */
  public assert(target: any, options?: { 
    /**
     * Fails the assertion in case the target has extra properties.
     */
    strict?: boolean;
    /**
     * Debug the validation process.
     */
    debug?: boolean;
  }): target is ModelInterface<this> {
    try {
      this.validate(target, {
        debug: !!options?.debug,
        throwErrors: true,
        strictTypes: !!options?.strict,
      });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Creates an object that matches the model.
   * Useful for creating mocks.
   */
  public createSample<T extends t.TypeOf<Schema>>(sample: T) {
    return sample;
  }

  /**
   * Extend one of the declared samples in the model declaration.
   * Useful for tests.
   */
  public extendSample<T extends Partial<t.TypeOf<Schema>>, K extends SampleKey>(base: K, sample: T) {
    return deepMerge(this.samples[base], sample);
  }
}

export type ModelInterface<T extends Model<any>> = t.TypeOf<T['base']>;
