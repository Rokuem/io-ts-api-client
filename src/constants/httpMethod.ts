export enum HttpMethod {
  /**
   * The `CONNECT` method establishes a tunnel to the server identified by the
   * target resource.
   */
  CONNECT = 'CONNECT',

  /**
   * The `DELETE` method deletes the specified resource.
   */
  DELETE = 'DELETE',

  /**
   * The `GET` method requests a representation of the specified resource.
   * Requests using GET should only retrieve data.
   */
  GET = 'GET',

  /**
   * The `HEAD` method asks for a response identical to that of a GET request,
   * but without the response body.
   */
  HEAD = 'HEAD',

  /**
   * The `OPTIONS` method is used to describe the communication options for the
   * target resource.
   */
  OPTIONS = 'OPTIONS',

  /**
   * The PATCH method is used to apply partial modifications to a resource.
   */
  PATCH = 'PATCH',

  /**
   * The `POST` method is used to submit an entity to the specified resource,
   * often causing a change in state or side effects on the server.
   */
  POST = 'POST',

  /**
   * The `PUT` method replaces all current representations of the target
   * resource with the request payload.
   */
  PUT = 'PUT',

  /**
   * The `TRACE` method performs a message loop-back test along the path to the
   * target resource.
   */
  TRACE = 'TRACE',

  /**

The LINK method is used to establish one or more relationships between an existing resource identified by the effective request URI and other resources. Metadata contained within Link header fields [RFC5988] provide the information about which other resources are being connected to the target resource and the type of relationship being established. A payload within a LINK request message has no defined semantics.

The semantics of the LINK method change to a "conditional LINK" if the request message includes an If-Modified-Since, If-Unmodified-Since, If-Match, If-None-Match, or If-Range header field ([I-D.ietf-httpbis-p4-conditional]). A conditional LINK requests that the relationship be established only under the circumstances described by the conditional header field(s).

LINK request messages are idempotent. For any pair of resources, only a single relationship of any given type can exist. However, multiple relationships of different types can be established between the same pair of resources.

LINK request messages are not safe, however, in that establishing a relationship causes an inherent change to the state of the target resource.

Responses to LINK requests are not cacheable. If a LINK request passes through a cache that has one or more stored responses for the effective request URI, those stored responses will be invalidated (see Section 6 of [I-D.ietf-httpbis-p6-cache]).

A single LINK request message can contain multiple Link header fields, each of which establishes a separate relationship with the target resource. In such cases, the server MUST accept the entire set of relationships atomically. If any of the specified relationships cannot be created, the server MUST NOT create any of them.

A successful response to a Link request that results in either the creation or modification of a relationship SHOULD be 200 (OK) if the response includes a representation describing the status, 201 (Created) if the action results in the creation of a new resource that represents the newly established relationship, 202 (Accepted) if the action has not yet been enacted, or 204 (No Content) if the action has been enacted but the response does not include a representation.

The LINK method MAY be overridden by human intervention (or other means) on the origin server. The client cannot be guaranteed that the operation has been carried out, even if the status code returned from the origin server indicates that the action has been completed successfully. However, the server SHOULD NOT indicate success unless, at the time the response is given, it intends to create or update the specified relationships.

If the LINK request message attempts to create or update an existing relationship and the server does not intend to comply with the request for any reason other than a client or server error, the server can return a 304 (Not Modified) response to indicate that no modifications have been made.
   */
  LINK = 'LINK',
  /**
   * The UNLINK method is used to remove one or more relationships between the existing resource identified by the effective request URI and other resources. Metadata contained within Link header fields [RFC5988] provide the information about the resources to which relationships of a specific type are to be removed. A payload within an UNLINK request message has no defined semantics.

The semantics of the UNLINK method change to a "conditional UNLINK" if the request message includes an If-Modified-Since, If-Unmodified-Since, If-Match, If-None-Match, or If-Range header field ([I-D.ietf-httpbis-p4-conditional]). A conditional UNLINK requests that the relationship be removed only under the circumstances described by the conditional header field(s).

UNLINK request messages are idempotent.

UNLINK request messages are not safe, however, in that removing a relationship causes an inherent change to the state of the target resource.

Responses to UNLINK requests are not cacheable. If an UNLINK request passes through a cache that has one or more stored responses for the effective request URI, those stored responses will be invalidated (see Section 6 of [I-D.ietf-httpbis-p6-cache]).

A single UNLINK request message can contain multiple Link header fields, each of which identifies a separate relationship to remove. In such cases, the server MUST remove the entire set of relationships atomically. If any of the specified relationships cannot be removed, the server MUST NOT remove any of them.

A successful response indicating the removing of the relationship SHOULD be 200 (OK) if the response includes a representation describing the status, 202 (Accepted) if the action has not yet been enacted, or 204 (No Content) if the action has been enacted but the response does not include a representation.

The UNLINK method MAY be overridden by human intervention (or other means) on the origin server. The client cannot be guaranteed that the operation has been carried out, even if the status code returned from the origin server indicates that the action has been completed successfully. However, the server SHOULD NOT indicate success unless, at the time the response is given, it intends to remove the specified relationships.

If the UNLINK request message attempts to remove an existing relationship and the server does not intend to remove or otherwise alter the existing relationship for any reason other than a client or server error, the server can return a 304 (Not Modified) response to indicate that no modifications have been made.
   */
  UNLINK = 'UNLINK',
}
