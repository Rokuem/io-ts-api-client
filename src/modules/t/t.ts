import * as types from 'io-ts';
import * as extraTypes from 'io-ts-types';

export const t = {
  ...types,
  ...extraTypes,
  /**
   * Creates null intersection with provided type.
   */
  nullable: <T extends types.Any>(type: T) => types.union([t.null, type]),
  /**
   * Creates intersection between required and optional props.
   * @since 0.7.4
   */
  schema: <T extends types.Props, P extends types.Props>(type: { required: T, optional: P }) => types.intersection([
    types.interface(type.required),
    types.partial(type.optional)
  ]),
  /**
   * Type for valid date strings.
   * @since 0.8.5
   */
  dateString: new types.Type<string, string, unknown>(
    'date string',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }

      const date = new Date(input);

      if (isNaN(date.getTime())) {
        return types.failure(input, context);
      }

      return types.success(input);
    },
    types.identity
  ),
  /**
   * Type for valid urls.
   * @since 0.8.5
   */
  url: new types.Type<string, string, unknown>(
    'url',
    (input: unknown): input is string => typeof input === 'string',
    (input, context) => {
      if (typeof input !== 'string') {
        return types.failure(input, context);
      }
  
      try {
        // URL constructors fails with invalid urls.
        new URL(input);
        return types.success(input);
      } catch (e) {
        return types.failure(input, context);
      }
    },
    types.identity
  )
};