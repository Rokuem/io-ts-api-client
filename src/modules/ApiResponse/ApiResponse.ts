import { Model } from '../Model/Model';
import { HttpStatus } from '../../constants/HttpStatus';

/**
 * Describes a possible response for the resource.
 */
export class ApiResponse<M extends Model, S extends HttpStatus> {
  public model!: M;
  public status!: S;
  constructor(options: { model: M; status: S }) {
    Object.assign(this, options);
  }
}
