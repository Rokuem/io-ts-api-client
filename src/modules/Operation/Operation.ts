import { AxiosError } from "axios";
import { HttpMethod } from "../../constants/httpMethod";
import { ApiResponse } from "../ApiResponse/ApiResponse";
import { Model } from "../Model/Model";
import { ModelInterface } from "../Model/Model";
import { ResponseWithStatus } from "../../helpers/ResponseWithStatus.type";
import { t } from "../t/t";
import { Client } from "../Client/Client";
import { Resource } from "../Resource/Resource";
import { ValidationOptions } from "../types";

/**
 * Operations describes a request.
 *
 * You can pass its options as the first type argument.
 */
export class Operation<
  Payload extends Model<any>,
  Responses extends [...ApiResponse<any, any>[]],
  Method extends HttpMethod,
  Options = never
> {
  /**
   * Name of the operation. Infered when the client is made.
   */
  public name = "Operation";
  /**
   * function to construct the url of the operation.
   *
   * @param url - The base URL object of the api.
   * @param options - The options of the operation.
   */
  public url!: (...args: [url: URL, options: Options]) => URL;
  /**
   * HTTP method of the operation.
   */
  public method!: Method;
  /**
   * IO-TS Model used to validate the payload.
   */
  public payloadModel?: Payload;
  /**
   * List of possible responses for the operation.
   *
   * This can be merged with global responses of a client.
   */
  public readonly responses!: [...Responses];
  /**
   * Options of the operation.
   *
   * This is just for typing so you can use `{} as Options` for example.
   * You can also provide this in the operations constructor.
   */
  public options?: Options;
  /**
   * Function to get the object of headers to send with the request.
   */
  public headers?:
    | Record<string, string>
    | ((options?: Options) => Record<string, string>);
  /**
   * Function to mock the response. Useful for testing and for development when the API is not ready.
   */
  public mock?: (
    /**
     * Computed url returned by the Operation url option.
     */
    url?: URL,
    /**
     * Payload returned by the PayloadConstructor
     */
    payload?: ModelInterface<Payload>
  ) => Promise<
    | Partial<
        {
          [key in keyof Responses]: Responses[key] extends ApiResponse<any, any>
            ? Partial<
                ResponseWithStatus<
                  Responses[key]["status"],
                  ModelInterface<Responses[key]["model"]>
                >
              >
            : Responses[key];
        }[number]
      >
    | false
  >;
  /**
   * Function to construct the payload using the provided options for the operation.
   */
  payloadConstructor?: (options?: Options) => ModelInterface<Payload>;
  constructor(
    config: Pick<
      Operation<Payload, Responses, Method, Options>,
      | "url"
      | "method"
      | "payloadModel"
      | "payloadConstructor"
      | "responses"
      | "options"
      | "headers"
      | "mock"
    >
  ) {
    Object.assign(this, config);
  }

  public setName(name: string) {
    this.name = name;

    for (const response of this.responses) {
      if (response.model instanceof Model) {
        response.model.operation = this.name;
      }
    }

    if (this.payloadModel) {
      this.payloadModel.operation = this.name;
    }
  }

  /**
   * Executes the operation using the provided configuration.
   */
  public async execute<GlobalResponses extends ApiResponse<any, any>>({
    axiosInstance,
    options,
    base,
    globalResponses,
    debug,
    throwErrors,
    strictTypes,
    globalHeaders,
  }: {
    globalHeaders: Record<string, string>;
  } & Pick<
    Client<
      Record<string, Resource<Record<string, this>, GlobalResponses>>,
      GlobalResponses
    >,
    "base" | "globalResponses" | "axiosInstance"
  > &
    ValidationOptions &
    Pick<this, "options">) {
    const log = (...message: any[]) => {
      if (debug) {
        console.log(...message);
      }
    };

    const logError = (...message: any[]) => {
      if (debug) {
        console.error(...message);
      }
    };

    if (this.payloadConstructor) {
      const payload = this.payloadConstructor?.(options);
      log("Payload: ", payload);
      this.payloadModel?.validate(payload, {
        debug,
        throwErrors,
        strictTypes,
      });
    }

    type GetResponses<T extends [...ApiResponse<any, any>[]]> = {
      [key in keyof T]: T[key] extends ApiResponse<any, any>
        ? ResponseWithStatus<T[key]["status"], ModelInterface<T[key]["model"]>>
        : T[key];
    }[number];

    type $Response =
      | GetResponses<Responses>
      | GetResponses<[...GlobalResponses[]]>;

    const requestUrl = this.url(base as any, options as any);

    if (this.mock) {
      log("mock detected!");
      const mock = await this.mock(
        requestUrl,
        this.payloadConstructor?.(options)
      );
      log("Mock: ", mock);
      if (mock !== false) {
        return { ...mock, body: (mock as any)?.data } as $Response;
      }
    }

    try {
      log("Making axios request");
      const response = await axiosInstance.request<$Response>({
        method: this.method as any,
        data: this.payloadConstructor?.(options),
        url: requestUrl.href,
        headers: {
          ...globalHeaders,
          ...(typeof this.headers === "function"
            ? this.headers?.(options)
            : this.headers || {}),
        },
        validateStatus: (status) =>
          this.responses.some((res) => res.status === status),
      });

      if (!response) {
        throw new Error(`[${this.name}] Received empty reponse`);
      }

      const matchingResponses = [...this.responses, ...globalResponses].filter(
        (res) => res.status === response.status
      );

      log("matching Responses: ", matchingResponses.length);

      let [responseDeclaration] = matchingResponses;

      if (matchingResponses.length > 1) {
        responseDeclaration = new ApiResponse({
          model: new Model({
            schema: t.intersection([
              ...matchingResponses.map((res) => res.model.base),
            ] as any),
            name: matchingResponses.map((res) => res.model.name).join(" | "),
          }),
          status: responseDeclaration.status,
        });
      }

      if (
        !responseDeclaration ||
        !(responseDeclaration.model instanceof Model)
      ) {
        console.error(`[${this.name}] UNEXPECTED RESPONSE: `, response);
        throw new Error(
          `Unexpected response without declaration. Status: ${
            response.status
          }, data: ${JSON.stringify(response.data)}`
        );
      }

      responseDeclaration.model.operation = this.name;

      const data = responseDeclaration.model.validate((response as any).data, {
        debug,
        strictTypes,
        throwErrors,
      });

      return {
        ...response,
        data,
        body: data,
      };
    } catch (error) {
      if (typeof error === "object" && error && "response" in error) {
        const axiosError = error as AxiosError;
        logError("Unexpected response: ", axiosError.response);
        axiosError.message = `MODEL NOT FOUND - Unexpected response received from operation ${
          this.name
        }: ${axiosError.response?.status} - ${JSON.stringify(
          axiosError.response?.data
        )}`;
      }

      throw error;
    }
  }
}
