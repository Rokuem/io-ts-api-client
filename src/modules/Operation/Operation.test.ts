import { Operation } from './Operation';
import { HttpMethod } from '../../constants/httpMethod';
import { t } from '../t/t';
import { Model } from '../Model/Model';
import { HttpStatus } from '../../constants/HttpStatus';
import { expectTypeOf } from 'expect-type';
import { ApiResponse } from '../ApiResponse/ApiResponse';

const operation = new Operation({
  url: (url) => url,
  method: HttpMethod.GET,
  payloadModel: new Model({
    name: 'operation payload',
    schema: t.undefined,
  }),
  responses: [
    new ApiResponse({
      model: new Model({
        name: 'bad response',
        schema: t.undefined,
      }),
      status: HttpStatus.OK,
    }),
    new ApiResponse({
      model: new Model({
        name: 'repeated response',
        schema: t.undefined,
      }),
      status: HttpStatus.OK,
    }),
    new ApiResponse({
      model: new Model({
        name: 'good response',
        schema: t.interface({
          ok: t.boolean,
        }),
      }),
      status: HttpStatus.ACCEPTED,
    }),
  ],
});

describe('Operation', () => {
  test('Should match the snapshot', () => {
    expect(operation).toMatchSnapshot();
  });

  test('Should infer The responses correctly', () => {
    expectTypeOf(operation.responses[0].status).toMatchTypeOf(
      HttpStatus.OK as const
    );
  });
});
