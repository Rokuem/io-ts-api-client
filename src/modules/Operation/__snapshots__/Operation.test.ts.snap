// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation Should match the snapshot 1`] = `
Operation {
  "method": "GET",
  "name": "Operation",
  "payloadModel": Model {
    "base": UndefinedType {
      "_tag": "UndefinedType",
      "decode": [Function],
      "encode": [Function],
      "is": [Function],
      "name": "undefined",
      "validate": [Function],
    },
    "name": "operation payload",
    "operation": "",
  },
  "responses": Array [
    ApiResponse {
      "model": Model {
        "base": UndefinedType {
          "_tag": "UndefinedType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "undefined",
          "validate": [Function],
        },
        "name": "bad response",
        "operation": "",
      },
      "status": 200,
    },
    ApiResponse {
      "model": Model {
        "base": UndefinedType {
          "_tag": "UndefinedType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "undefined",
          "validate": [Function],
        },
        "name": "repeated response",
        "operation": "",
      },
      "status": 200,
    },
    ApiResponse {
      "model": Model {
        "base": InterfaceType {
          "_tag": "InterfaceType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "{ ok: boolean }",
          "props": Object {
            "ok": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
        "name": "good response",
        "operation": "",
      },
      "status": 202,
    },
  ],
  "url": [Function],
}
`;
